portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,94 @@
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.data.Task;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Task").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Task> tasks = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String comment = (String) entity.getProperty("comment");
+      long timestamp = (long) entity.getProperty("timestamp");
+      
+      Task task = new Task(id, comment, timestamp);
+      tasks.add(task);
+    }
+
+    int bounded_task = getIntParameter(request, response, "comment-bound", 0);
+
+    List tasks_ = tasks.subList(0, Math.min(tasks.size(), bounded_task));
+
+    response.setContentType("application/json");
+    String json = new Gson().toJson(tasks_);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String text = getParameter(request, "text-input", "");
+    String username = getParameter(request, "name-input", "Anonymous");
+    long timestamp = System.currentTimeMillis();
+
+    String comment = username + ": " + text;
+
+    // Add Comments
+    Entity taskEntity = new Entity("Task");
+    taskEntity.setProperty("comment", comment);
+    taskEntity.setProperty("timestamp", timestamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(taskEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  // Helper Functions
+  private String getParameter(HttpServletRequest request, String comment, String defaultValue) {
+    String value = request.getParameter(comment);
+    if (value == null || "".equals(value)) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  private int getIntParameter(HttpServletRequest request, HttpServletResponse response, String comment_bound, int defaultValue) throws IOException {
```

Comment:
In order to make this as generally useful as possible, you dont want to name the parameters for one particular case, but for just the general case. So instead of `comment_bound` this could be `parameter_name`